### PR TITLE
Activate asserts

### PR DESCRIPTION
**Pull request by armintoepfer**
We can use 

``` sh
CMAKE_BUILD_TARGET=Debug pip install --verbose --upgrade --no-deps .
```

to install a debug version of CC2 and run `make check` of GC with active asserts

### CODE REVIEW

**Comment by dalexander** [comment]
In my experience "Debug" is criminally slow, is there a build mode where it optimizes but still retains asserts?  I've often seen "fastdebug" modes in other systems

**Comment by dalexander** [comment]
Armin can you rebase this?  I'm not OK with us merging things into master that introduce (multiple!) silly merge commits.

**Comment by dalexander** [comment]
thanks for the rebase.  Regarding the "fastdebug", is there any such possibility?  If not, I will merge as is.

**Comment by armintoepfer** [comment]
I added CMAKE_BUILD_TYPE=ReleaseWithAssert, if I understood correctly that is what you wanted?

**Comment by dalexander** [comment]
This looks good to me but I want to ask @nlhepler if it looks good to him, since he has been more involved in the build system than me lately.

File: cmake/uny-compilerflags.cmake [hunk 1]
```diff
@@ -40,4 +40,34 @@ if (SANITIZE)
...
+    "Flags used for linking binaries during ReleaseWithAssert builds."
+    FORCE)
+SET(CMAKE_SHARED_LINKER_FLAGS_RELEASEWITHASSERT
+    "${CMAKE_SHARED_LINKER_FLAGS}" CACHE STRING
+    "Flags used by the shared libraries linker during ReleaseWithAssert builds."
+    FORCE)
+MARK_AS_ADVANCED(
+    CMAKE_CXX_FLAGS_RELEASEWITHASSERT
+    CMAKE_EXE_LINKER_FLAGS_RELEASEWITHASSERT
+    CMAKE_SHARED_LINKER_FLAGS_RELEASEWITHASSERT)
```
Comment (nlhepler) [inline cmake/uny-compilerflags.cmake]: can we move these to a uny-releaseWithAssert.cmake under cmake/?

File: cmake/uny-compilerflags.cmake [hunk 2]
```diff
@@ -40,4 +40,34 @@ if (SANITIZE)
 
     # UndefinedBehaviorSanitizer is a fast undefined behavior detector.
     set(UNY_DEBUG_FLAGS "${UNY_DEBUG_FLAGS} -fsanitize=undefined")
+endif()
+
+# shared CXX flags for src & tests
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${UNY_FLAGS}")
+set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} ${UNY_DEBUG_FLAGS}")
+SET(CMAKE_EXE_LINKER_FLAGS  "${CMAKE_EXE_LINKER_FLAGS} ${UNY_LINKER_FLAGS}")
```
Comment (nlhepler) [inline cmake/uny-compilerflags.cmake]: These are fine here, but should we probably start using:

set_target_properties($TARGET PROPERTIES
    COMPILE_FLAGS "${UNY_FLAGS}"
    LINK_FLAGS "${UNY_LINKER_FLAGS}"
)

?

File: cmake/uny-compilerflags.cmake (hunk 1 above)
Comment (nlhepler) [inline cmake/uny-compilerflags.cmake]: Alternatively, theres: https://github.com/aalhossary/gromacs-HREMD/blob/master/cmake/gmxBuildTypeReleaseWithAssert.cmake

File: cmake/uny-compilerflags.cmake (hunk 2 above)
Comment (armintoepfer) [inline cmake/uny-compilerflags.cmake]: But we'd have to set it for both targets individually?

File: cmake/uny-compilerflags.cmake (hunk 1 above)
Comment (armintoepfer) [inline cmake/uny-compilerflags.cmake]: I like that, added

**Comment by nlhepler** [comment]
this looks good to me =)

**Pull request merged**